### PR TITLE
feat: Add Equals function to Dictionary (for future de-duplication of ship.baseAttributes)

### PR DESCRIPTION
**Pull request by petervdmeer**
**Feature:** This PR implements a first step towards re-using Ship.baseAttributes among ships.

## Feature Details
The `Ship.baseAttributes` will be identical for most ships of the same Model, but currently every ship in the game contains a copy of all the baseAttributes. Active ships in the game also contain an `const Outfit *explosionWeapon` pointer to the baseAttributes of their Models in GameData to ensure that explosions can be properly shown without pointer-exceptions after the ship-instance dies.

This PR is about adding Equals functions to relevant classes (at initial upload only to the Dictionary class, but I expect Weapon and Outfit to also get an Equals function in this PR or in follow-up PRs). The Equals function should help to centrally store the BaseAttributes without duplication and with the ability to keep a stable pointer to BaseAttributes from the ship (even after it dies, gets removed and still is exploding).

## UI Screenshots /  Usage Examples
N/A

## Testing Done
I added an unit-test for the Equals function in Dictionary.

## Performance Impact
No impact expected in the RealTime section of the game; BaseAttributes don't seem to be used in the RealTime sections of the game, since there the total set of Attributes including the data from installed outfits seems to be relevant.

### CODE REVIEW

File: source/Dictionary.h [hunk 1]
```diff
@@ -32,6 +32,9 @@ class Dictionary : private std::vector<std::pair<const char *, double>> {
 	double Get(const char *key) const;
 	double Get(const std::string &key) const;
 	
+	// Compare two dictionaries to see if they contain exactly the same data.
+	bool Equals(Dictionary &other);
```
Comment (tehhowch) [inline source/Dictionary.h:36]:
```suggestion
	bool Equals(const Dictionary &other) const noexcept;
```

**Review by tehhowch — COMMENTED** [review]
Sorry, but I don't really see the purpose of this method at the moment.
While currently no ship can have a custom explosion, that is something I (and others) would like to change.

For example, destroying a ship using some arbitrary weapon might result in a differently colored explosion, or one with mutated characteristics. To reach that scenario, we will need to pass the `Weapon` from a ship instance by value, rather than a pointer to a base instance.
(This will also allow us to restore the "hidden" nature of many person ships, as their custom ship models will no longer need to be declared as root-level nodes in order to have a non-empty explosion.)


Changes to how we instantiate ship instances should result in the removal of the "baseAttributes" field, too, as there will be a distinction between a Ship instance and a Ship "template".

File: source/Dictionary.h (hunk 1 above)
Comment (petervdmeer) [reply source/Dictionary.h]: Yes, should be const. I'll update this in the next commit.

**Comment by petervdmeer** [comment]
> Sorry, but I don't really see the purpose of this method at the moment.

No problem. I expect this function to be useful when we want to to de-duplication when we automatically translate "full ships" into "ship templates" and "ship instances". But we didn't even decide if we want to do such automatic translation and if we then also didn't decide if we want to do de-duplication there, so I guess this PR is a bit premature.

> While currently no ship can have a custom explosion, that is something I (and others) would like to change.

I am also a member of this "others" group. :smiley:
  
> For example, destroying a ship using some arbitrary weapon might result in a differently colored explosion, or one with mutated characteristics. To reach that scenario, we will need to pass the `Weapon` from a ship instance by value, rather than a pointer to a base instance.

Yes, passing the "Weapon" by value is one of the options I considered, but I also came up with other options:
- Keeping the ship object alive until after the "Weapon" finishes. (So you can just pass a pointer to the Weapon in the ship itself.)
- Storing the "Weapon" itself outside of the Ship classes (in GameData) so that it can stay alive after the ship dies. (This is the option that this PR relates to.)

> Changes to how we instantiate ship instances should result in the removal of the "baseAttributes" field, too, as there will be a distinction between a Ship instance and a Ship "template".

I agree and I had some similar ideas. When we go this route then we will also need to put some thought into dealing with existing savegames, since those savegames currently contain mostly ship "templates", while they should contain "ship instances" in the future. I listed some options on how we could deal with this in the "additional context" part of https://github.com/endless-sky/endless-sky/issues/4654 (where chassis can be read as "template" and "active ship" as "ship instance").

**Pull request closed without merging**